### PR TITLE
[Tests] Ledger internals property tests

### PR DESCRIPTION
**Pull request by dkijania**

### CODE REVIEW

File: chain-impl-mockchain/src/accounting/account/mod.rs [hunk 1]
```diff
@@ -490,4 +414,69 @@ pub mod tests {
...
+    pub fn ledger_removes_account_only_if_zeroed(
+        id: Identifier,
+        account_state: AccountState<()>,
+    ) -> TestResult {
+        let mut ledger = Ledger::new();
+        ledger = ledger
+            .add_account(&id, account_state.get_value(), ())
+            .unwrap();
+        let result = ledger.remove_account(&id);
+        let expected_result = account_state.get_value() == Value::zero();
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/accounting/account/mod.rs]: I would name this boolean `expected_zero` for clarity.

File: chain-impl-mockchain/src/accounting/account/mod.rs [hunk 2]
```diff
@@ -490,4 +414,69 @@ pub mod tests {
...
+            (Ok(_), false) => TestResult::failed(),
+            (Err(_), true) => TestResult::failed(),
+            (Ok(ledger), true) => verify_account_does_not_exist(&ledger, &id),
+        }
+    }
+
+    fn verify_account_exists(ledger: &Ledger, id: &Identifier) -> TestResult {
+        match ledger.exists(&id) {
+            true => TestResult::passed(),
+            false => TestResult::error(format!("Account ({:?}) not exists , while it should", &id)),
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/accounting/account/mod.rs]: Spelling: "does not exist, " (mind the comma spacing too)

File: chain-impl-mockchain/src/accounting/account/mod.rs [hunk 3]
```diff
@@ -490,4 +414,69 @@ pub mod tests {
...
+        }
+    }
+
+    fn verify_account_does_not_exist(ledger: &Ledger, id: &Identifier) -> TestResult {
+        match ledger.exists(&id) {
+            true => TestResult::error(format!("Account ({:?}) exists , while it should not", &id)),
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/accounting/account/mod.rs]: Space before comma, grrr :)

File: chain-impl-mockchain/src/ledger/iter.rs [hunk 4]
```diff
@@ -240,3 +240,110 @@ impl<'a> std::iter::FromIterator<Entry<'a>> for Result<Ledger, Error> {
...
+        ));
+        let (_block0_hash, ledger) =
+            ledger::create_initial_fake_ledger(&[message], ConfigBuilder::new().build()).unwrap();
+
+        // FIXME: generate arbitrary ledger
+
+        for item in ledger.iter() {
+            match item {
+                Entry::Globals(globals) => {
+                    println!(
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/ledger/iter.rs]: Is it normal to use `println` in these tests?
This looks like it needs rustfmt, too.

File: chain-impl-mockchain/src/ledger/ledger.rs [hunk 5]
```diff
@@ -1015,3 +1014,1188 @@ fn input_multi_account_verify<'a>(
...
+        }
+    }
+
+    #[quickcheck]
+    pub fn match_identifier_witness_prop_test(
+        id: AccountIdentifier,
+        witness: Witness,
+    ) -> TestResult {
+        let result = super::match_identifier_witness(&id, &witness);
+        match (witness.clone(), result) {
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/ledger/ledger.rs]: `clone` should not be necessary, but if it saves some syntax pain, it's OK for test code.

File: chain-impl-mockchain/src/ledger/ledger.rs [hunk 6]
```diff
@@ -1015,3 +1014,1188 @@ fn input_multi_account_verify<'a>(
...
+                false => {
+                    let utxo = utxo.unwrap();
+                    assert_eq!(utxo.output.value, value);
+                    return self;
+                }
+            }
+        }
+    }
+
+    fn verify_that(
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/ledger/ledger.rs]: Uninformative method name, the one below as well.

File: chain-impl-mockchain/src/ledger/iter.rs (hunk 4 above)
Comment (dkijania) [reply chain-impl-mockchain/src/ledger/iter.rs]: I don't really know, just move this test into new location.

File: chain-impl-mockchain/src/ledger/ledger.rs [hunk 7]
```diff
@@ -1404,19 +1404,23 @@ mod tests {
         }
     }
 
-    pub struct InternalApplyTransactionVerifier {
+    pub struct LedgerStateVerifier {
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/ledger/ledger.rs]: Good, I was concerned about these "internal" names.

File: chain-impl-mockchain/src/testing/data/address.rs [hunk 8]
```diff
@@ -95,20 +93,46 @@ impl AddressData {
...
                 Input::from_account_public_key(self.public_key(), value.clone())
             }
-            Kind::Single { .. } | Kind::Group { .. } | Kind::Multisig { .. } => {
+            Kind::Single { .. } | Kind::Group { .. } => {
                 Input::from_utxo_entry(utxo.expect(&format!(
                     "invalid state, utxo should be Some if Kind not Account {:?}",
                     &self.address
                 )))
             }
+            Kind::Multisig { .. } => panic!("unimplemented"),
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/testing/data/address.rs]: There is `unimplemented!()` macro just for this purpose.

File: chain-impl-mockchain/src/testing/arbitrary/mod.rs [hunk 9]
```diff
@@ -63,4 +66,4 @@ impl Arbitrary for Output<Address> {
             value: Arbitrary::arbitrary(g),
         }
     }
-}
+}
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/testing/arbitrary/mod.rs]: Nitpick: missing newline. Rustfmt?

File: chain-impl-mockchain/src/testing/data/leader.rs [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+#[derive(Clone)]
+pub struct LeaderPair {
+    leader_id: LeaderId,
+    leader_key: SecretKey<Ed25519Extended>,
+}
+
+impl PartialEq<LeaderPair> for LeaderPair {
+    fn eq(&self, other: &LeaderPair) -> bool {
+        self.id() == other.id()
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/testing/data/leader.rs]: This looks very dodgy as a `PartialEq` impl.
There is nothing that prevents construction of pairs with the same ID but different keys, so I'd rather remove this and compare the IDs directly where needed.

File: chain-impl-mockchain/src/testing/extensions/mod.rs [hunk 11]
```diff
@@ -0,0 +1,17 @@
+use crate::{accounting::account::AccountState, value::Value};
+
+impl AccountState<()> {
```
Comment (mzabaluev) [inline chain-impl-mockchain/src/testing/extensions/mod.rs]: Maybe use a different name than the type under test to reduce ambiguity. `AccountStateValidator`?

Comment (dkijania) [reply chain-impl-mockchain/src/testing/extensions/mod.rs]: Actually it's not needed anymore

**Pull request merged**